### ISSUE TITLE
[SOLVED] Graphical server shuts down immediately after starting on macOS Catalina

### ISSUE THREAD

**Issue by andrewsonin** [merenlab/anvio#1274]
Hi there,
I installed anvio-6 as follows:
```
conda create -n anvio python=3.6
conda activate anvio
conda config --env --add channels conda-forge
conda config --env --add channels bioconda
conda install -y anvio=6
conda install -y diamond=0.9.14
```

Then I run `anvi-self-test --suite mini` and it ends without any error. The main problem is that it ends immediately after calling a web-browser to create an interactive page.

[Here](https://drive.google.com/file/d/1W_NavqHZ-_RVK0AFWqrPowUazE52cs9j/view?usp=sharing) is what the web-browser looks like. The error is reproduced in both Safari and Chrome. It is also reproduced for other anvio interactive programs (e.g. `anvi-display-pan`).

Below is the result of calling `anvi-self-test --version`:
```
Anvi'o version ...............................: esther (v6)
Profile DB version ...........................: 31
Contigs DB version ...........................: 14
Pan DB version ...............................: 13
Genome data storage version ..................: 6
Auxiliary data storage version ...............: 2
Structure DB version .........................: 1
```
And here is the output of `conda list`:
```
# packages in environment at /Users/andrewsonin/.programs/anaconda3/envs/anvio:
#
# Name                    Version                   Build  Channel
anvio                     6                             0    bioconda
anvio-minimal             6                          py_0    bioconda
appdirs                   1.4.3                      py_1    conda-forge
asn1crypto                1.2.0                    py36_0    conda-forge
attrs                     19.3.0                     py_0    conda-forge
bcftools                  1.9                  h16e57c4_7    bioconda
biopython                 1.74             py36h01d97ff_0    conda-forge
blast                     2.6.0               boost1.64_2    bioconda
boost                     1.70.0           py36haf112f3_1    conda-forge
boost-cpp                 1.70.0               h75728bb_2    conda-forge
bottle                    0.12.13                    py_1    conda-forge
bowtie2                   2.3.5            py36h5c9b4e4_0    bioconda
bwa                       0.7.17               h2573ce8_6    bioconda
bzip2                     1.0.8                h01d97ff_1    conda-forge
ca-certificates           2019.9.11            hecc5488_0    conda-forge
centrifuge                1.0.4_beta      py36pl526ha92aebf_2    bioconda
certifi                   2019.6.16                py36_1    conda-forge
cffi                      1.13.0           py36h33e799b_0    conda-forge
chardet                   3.0.4                 py36_1003    conda-forge
cherrypy                  8.0.0                    py36_0    conda-forge
colored                   1.3.93                     py_0    conda-forge
configargparse            0.13.0                     py_1    conda-forge
cryptography              2.7              py36h212c5bf_0    conda-forge
curl                      7.62.0               ha441bb4_0  
cycler                    0.10.0                     py_1    conda-forge
cython                    0.29.13          py36h6de7cb9_0    conda-forge
datrie                    0.8              py36h01d97ff_0    conda-forge
dbus                      1.13.6               h2f22bb5_0    conda-forge
diamond                   0.9.14               hd28b015_2    bioconda
django                    2.0.8                    py36_0    conda-forge
docutils                  0.15.2                   py36_0    conda-forge
ete3                      3.1.1                    py36_0    bioconda
expat                     2.2.5             h6de7cb9_1003    conda-forge
freetype                  2.10.0               h24853df_1    conda-forge
gettext                   0.19.8.1          h46ab8bc_1002    conda-forge
glib                      2.58.3            h9d45998_1002    conda-forge
gsl                       2.5                  ha2d443c_1    conda-forge
hmmer                     3.2.1                h0a44026_1    bioconda
htslib                    1.9                  hd68223c_5    bioconda
icu                       64.2                 h6de7cb9_1    conda-forge
idna                      2.7                   py36_1002    conda-forge
illumina-utils            2.6                        py_0    bioconda
importlib_metadata        0.23                     py36_0    conda-forge
iqtree                    1.6.12               h2dec4b4_0    bioconda
joblib                    0.14.0                     py_0    conda-forge
jpeg                      9c                h1de35cc_1001    conda-forge
jsonschema                3.1.1                    py36_0    conda-forge
kiwisolver                1.1.0            py36h770b8ee_0    conda-forge
libblas                   3.8.0               14_openblas    conda-forge
libcblas                  3.8.0               14_openblas    conda-forge
libclang                  8.0.1                h770b8ee_1    conda-forge
libcurl                   7.62.0               h051b688_0  
libcxx                    4.0.1                hcfea43d_1  
libcxxabi                 4.0.1                hcfea43d_1  
libdeflate                1.0                  h1de35cc_1    bioconda
libedit                   3.1.20181209         hb402a30_0  
libffi                    3.2.1                h475c297_4  
libgfortran               4.0.0                         2    conda-forge
libiconv                  1.15              h01d97ff_1005    conda-forge
liblapack                 3.8.0               14_openblas    conda-forge
libllvm8                  8.0.1                h770b8ee_0    conda-forge
libopenblas               0.3.7                h4bb4525_2    conda-forge
libpng                    1.6.37               h2573ce8_0    conda-forge
libssh2                   1.8.2                hcdc9a53_2    conda-forge
libxcb                    1.13              h1de35cc_1002    conda-forge
libxml2                   2.9.9                h12c6b28_5    conda-forge
libxslt                   1.1.33               h320ff13_0    conda-forge
llvm-openmp               8.0.1                h770b8ee_0    conda-forge
lxml                      4.4.1            py36h08abf6f_0    conda-forge
matplotlib                2.2.3                    py36_1    conda-forge
matplotlib-base           2.2.3            py36hd64abcd_1    conda-forge
mcl                       14.137          pl526h470a237_4    bioconda
megahit                   1.2.9                hfbae3c0_0    bioconda
mistune                   0.8.1            py36h638d0ca_0  
more-itertools            7.2.0                      py_0    conda-forge
mummer                    3.23                          8    bioconda
muscle                    3.8.1551             h04f5b5a_4    bioconda
ncurses                   6.1                  h0a44026_1  
nose                      1.3.7                 py36_1002    conda-forge
nspr                      4.20              h0a44026_1000    conda-forge
nss                       3.46                 h39b4b1f_0    conda-forge
numpy                     1.17.2           py36h6b0580a_0    conda-forge
openmp                    8.0.1                         0    conda-forge
openssl                   1.1.1c               h01d97ff_0    conda-forge
pandas                    0.25.1           py36h86efe34_0    conda-forge
patsy                     0.5.1                      py_0    conda-forge
pcre                      8.41              h0a44026_1003    conda-forge
perl                      5.26.2            haec8ef5_1006    conda-forge
pip                       19.2.3                   py36_0  
prodigal                  2.6.3                         1    bioconda
psutil                    5.4.3                    py36_0    conda-forge
pthread-stubs             0.4               h1de35cc_1001    conda-forge
pyani                     0.2.7            py36h24bf2e0_1    bioconda
pycparser                 2.19                     py36_1    conda-forge
pyopenssl                 19.0.0                   py36_0    conda-forge
pyparsing                 2.4.2                      py_0    conda-forge
pyqt                      5.12.3           py36h2a560b1_0    conda-forge
pyqt5-sip                 4.19.18                  pypi_0    pypi
pyqtwebengine             5.12.1                   pypi_0    pypi
pyrsistent                0.15.4           py36h01d97ff_0    conda-forge
pysam                     0.15.2           py36haf89782_1    bioconda
pysocks                   1.7.1                    py36_0    conda-forge
python                    3.6.9                h359304d_0  
python-dateutil           2.8.0                      py_0    conda-forge
python-levenshtein        0.12.0                   pypi_0    pypi
pytz                      2019.3                     py_0    conda-forge
pyyaml                    5.1.2            py36h01d97ff_0    conda-forge
qt                        5.12.5               h1b46049_0    conda-forge
ratelimiter               1.2.0                 py36_1000    conda-forge
readline                  7.0                  h1de35cc_5  
requests                  2.20.0                py36_1000    conda-forge
samtools                  1.9                 h8aa4d43_12    bioconda
scikit-learn              0.21.2           py36hd4ffd6c_1    conda-forge
scipy                     1.3.1            py36h7e0e109_2    conda-forge
seaborn                   0.9.0                      py_1    conda-forge
setuptools                41.4.0                   py36_0  
six                       1.11.0                py36_1001    conda-forge
snakemake-minimal         5.2.4                    py36_0    bioconda
sqlite                    3.30.0               ha441bb4_0  
statsmodels               0.9.0           py36h917ab60_1000    conda-forge
tabulate                  0.8.3                      py_0    conda-forge
tbb                       2019.8               h770b8ee_0    conda-forge
tk                        8.6.8                ha441bb4_0  
tornado                   6.0.3            py36h01d97ff_0    conda-forge
trimal                    1.4.1                h04f5b5a_3    bioconda
urllib3                   1.23                  py36_1001    conda-forge
wheel                     0.33.6                   py36_0  
wrapt                     1.11.2           py36h01d97ff_0    conda-forge
xorg-kbproto              1.0.7             h1de35cc_1002    conda-forge
xorg-libice               1.0.10               h01d97ff_0    conda-forge
xorg-libsm                1.2.3             h01d97ff_1000    conda-forge
xorg-libx11               1.6.9                h0b31af3_0    conda-forge
xorg-libxau               1.0.9                h1de35cc_0    conda-forge
xorg-libxdmcp             1.1.3                h01d97ff_0    conda-forge
xorg-libxext              1.3.4                h01d97ff_0    conda-forge
xorg-libxrender           0.9.10            h01d97ff_1002    conda-forge
xorg-renderproto          0.11.1            h1de35cc_1002    conda-forge
xorg-xextproto            7.3.0             h1de35cc_1002    conda-forge
xorg-xproto               7.0.31            h1de35cc_1007    conda-forge
xz                        5.2.4                h1de35cc_4  
yaml                      0.1.7             h1de35cc_1001    conda-forge
zipp                      0.6.0                      py_0    conda-forge
zlib                      1.2.11               h1de35cc_3
```

**Comment by meren** [comment]
Just for clarity, do you see an error on your terminal?

**Comment by meren** [comment]
And what is it you see in when you open the developer console? :)

**Comment by andrewsonin** [comment]
> Just for clarity, do you see an error on your terminal?

In short, no.
I ran `anvi-self-test --suite mini > anvi-self-test_suite_mini.log 2>&1` and the result is [here](https://drive.google.com/open?id=1ueeS3EQZbDUP-TsKu7u3PEWCVh-GbvRW). Open it not with `vim`, but with `cat`.

> And what is it you see in when you open the developer console? :)

I see [this](https://drive.google.com/file/d/1Sk5UAGp36fsjj2Qsa8IjhhN5BsKs-_J4/view?usp=sharing).

**Comment by meren** [comment]
This is very disturbing. I don't see anything smoking in these outputs.

Can you please run `anvi-interactive` on one of your old profile databases that you have been able to visualize before? Perhaps the console (preferably Chrome) or the terminal window in which you will be running `anvi-interactive` would give you something useful in that case.

**Comment by meren** [comment]
There is this one we discovered recently and will address very soon:

https://github.com/merenlab/anvio/issues/1263

But yours doesn't have anything to do with that.

**Comment by andrewsonin** [comment]
[This](https://drive.google.com/file/d/1qvjqjUgeybTwpW6AYjjBAotmdkg4Mv9m/view?usp=sharing) is what I see in the terminal.

And [this](https://drive.google.com/file/d/19FoQW-9E2S9ChyQctDyl9vgJTVmcp-Qe/view?usp=sharing) is the Chrome console.

**Comment by meren** [comment]
This is absolutely crazy. It simply starts listening to the port successfully, and immediately dies.

We will have to look into this, @ozcan.

**Comment by meren** [comment]
Thank you for your help, @andrewsonin. If you manage to find out more, please let us know.

**Comment by meren** [comment]
Hey @andrewsonin, I think 51693f98b3564570d5fdd195acec685955bbabce solves it. If you are following `master`, can you please take a look? Otherwise you can wait a day or two as we will release a new version with this fix.

Best,

**Comment by meren** [comment]
I am sorry. I closed this by mistake. This in fact is not yet resolved :(

**Comment by ozcan** [comment]
Hi some people reported that using `sudo` with anvi'o commands solves this problem
However, I couldn't reproduce it myself as Catalina installer fails for me.

**Comment by andrewsonin** [comment]
@ozcan 
Yes, I tried and this works. Thank you!

**Comment by andrewsonin** [comment]
@ozcan 
I also noticed that after starting the interactive interface with `sudo`, subsequent program launches do not require superuser privileges to work correctly.

**Comment by andrewsonin** [comment]
@ozcan 
... and also [this](https://github.com/merenlab/anvio/issues/1215) error disappears.

**Comment by andrewsonin** [comment]
@ozcan 
Oh, just forget what I said in the comment above.

**Comment by lhor** [comment]
I'm having the same issue running macOS Catalina 10.15.2. The `sudo` command does the trick though. I don't think I have anything to add as everything else works fine but let me know if you need more information from my side.

**Comment by khendarg** [comment]
A bit of background: We're getting the same issue on macOS Catalina 10.15.1, but because our Macs are multi-user systems, our non-admins don't get to use the `sudo` workaround. For now, it seems to work on Ubuntu 18.04 and macOS High Sierra 10.13.

I did come close to tracking down what happens right before the server goes down with nary an exception and got to `socket.accept` calling `_socket._accept` (the source of which is at https://github.com/python/cpython/blob/master/Modules/socketmodule.c). The chain I went through was something like this:

 * anvi-interactive

 * anvio.bottleroutes.BottleApplication

 * bottle.Bottle

 * cherrypy.wsgiserver.HTTPServer.tick

 * _socket.socket.sock_accept

Unfortunately, I don't know much about server code or how to debug CPython modules. Is anyone familiar enough with Mojave->Catalina changes to pinpoint whom/where to forward this issue to?

**Comment by chloezhaoym** [comment]
Hi all,

I got the same issue when I run the interactive interface on Catalina 10.15.2 and sudo did solve my problem. But today the system upgraded to macOS Catalina 10.15.3 and sudo failed me this time. Can anyone help on this? Thanks!!

**Comment by chloezhaoym** [comment]
Hi all,

I got a very wired thing happened today after upgrade to macOS Catalina 10.15.3. 
When I try to use sudo to solve the grafical server shutdown issue, it worked for anvi-self-test command very well, but failed when I try to look at my contigs using command [anvi-interactive]. The graphical server shuts itself down again even if I put sudo in my command.
--the sudo trick did work fine on Catalina 10.15.2 for [anvi-interactive] earlier on my computer.

Can anyone help me on this? Thanks!!

**Comment by meren** [comment]
Hey @khendarg, I know it's been ages since you commented on this, but I finally have been able to find out why this is happening (I had to, because I installed anvi'o on a new computer that runs Mac OS Catalina, and anvi'o started crashing all the time).

In a nutshell, the problem is due to network requests. Whenever anvi'o runs this line in bottle routes module, the server gets killed:

```
news_markdown = requests.get('https://raw.githubusercontent.com/merenlab/anvio/master/NEWS.md')
```

It is the same with other ways to access to this file. It seems whenever we make a network request, the process is killed abruptly. It doesn't matter whether it is through http or https. The solution is to search for the word 'news.js' in this file:

``` bash
nano `python -c "import anvio; import os; print(os.path.dirname(anvio.__file__))"`/data/interactive/index.html
```

Remove that line, and save it.

I have no idea how else to solve this.
